block-producer/src/tasks/txqueue/mod.rs

```diff
@@ -0,0 +1,315 @@
+    queue_task.run().await
+}
+
+/// The transaction queue task
+#[derive(Clone)]
+struct TxQueue<HandleIn, HandleOut>
+where
+    HandleIn: TxQueueHandleIn,
+    HandleOut: TxQueueHandleOut,
+{
+    ready_queue: ReadyQueue,
+    handle_in: Arc<HandleIn>,
+    handle_out: Arc<HandleOut>,
+    options: TxQueueOptions,
+}
+
+impl<HandleIn, HandleOut> TxQueue<HandleIn, HandleOut>
+where
+    HandleIn: TxQueueHandleIn,
+    HandleOut: TxQueueHandleOut,
+{
+    pub fn new(
+        handle_in: HandleIn,
+        handle_out: HandleOut,
+        options: TxQueueOptions,
+    ) -> Self {
+        Self {
+            ready_queue: Arc::new(Mutex::new(Vec::new())),
+            handle_in: Arc::new(handle_in),
+            handle_out: Arc::new(handle_out),
+            options,
+        }
+    }
+
+    /// Start the task
+    pub async fn run(self) {
+        let tx_queue = Arc::new(self);
+        let timer_task_handle = start_timer_task(
+            tx_queue.ready_queue.clone(),
+            tx_queue.handle_out.clone(),
+            tx_queue.options.tx_max_time_in_queue,
+            tx_queue.options.batch_size,
+        );
+
+        loop {
+            // Handle new transaction coming in
+            let proven_tx =
+                tx_queue.handle_in.read_transaction().await.expect("Failed to read transaction");
+            let tx_queue = tx_queue.clone();
+            let timer_task_handle = timer_task_handle.clone();
+
+            tokio::spawn(async move {
+                tx_queue.on_read_transaction(proven_tx, timer_task_handle).await
+            });
+        }
+    }
+
+    // HELPERS
+    // --------------------------------------------------------------------------------------------
+
+    async fn on_read_transaction(
+        self: Arc<TxQueue<HandleIn, HandleOut>>,
+        proven_tx: ProvenTransaction,
+        timer_task_handle: TimerTaskHandle,
+    ) {
+        let proven_tx = Arc::new(proven_tx);
+
+        let verification_result = self
+            .handle_out
+            .verify_transaction(proven_tx.clone())
+            .await
+            .expect("Failed to verify transaction");
+
+        if let Err(_failure_reason) = verification_result {
+            // TODO: Log failure properly
+            return;
+        }
+
+        // Transaction verification succeeded. It is safe to add transaction to queue.
+        let mut locked_ready_queue = self.ready_queue.lock().await;
+
+        if locked_ready_queue.is_empty() {
+            timer_task_handle.start_timer();
+        }
+
+        locked_ready_queue.push(proven_tx);
+
+        if locked_ready_queue.len() >= self.options.batch_size {
+            // FIXME: Dropping the ready queue here means that 2 tasks could
+            // send a batch, where the first batch will contain `batch_size`
+            // transactions, and the second would contain only 1 transaction.
+            // This is low-risk and has little-to-no impact if it occurs.
+            drop(locked_ready_queue);
```

Comment:
Why do we need to drop the locked queue here? Would we be able to pass it into `send_batch()` and avoid the potential issue described in the comment?

Reply:
I was concerned with potential future deadlocks that arise by holding a mutex across a `tokio::spawn()`. I fixed it differently: `send_batch()`'s caller is responsible for draining the queue.

Fixed in [99542df](https://github.com/0xPolygonMiden/miden-node/pull/44/commits/99542df8374b4d255740e6f572a4e7d4a8a9a7ae)